daemon/emer-aggregate-tally.c

```diff
@@ -0,0 +1,391 @@
+
+struct _EmerAggregateTally
+{
+  GObject parent_instance;
+
+  gchar *persistent_cache_directory;
+};
+
+G_DEFINE_TYPE (EmerAggregateTally, emer_aggregate_tally, G_TYPE_OBJECT)
+
+enum {
+  PROP_0,
+  PROP_PERSISTENT_CACHE_DIRECTORY,
+  N_PROPS,
+};
+
+static GParamSpec *properties[N_PROPS] = { NULL, };
+
+static void
+ensure_folder_for_date (EmerAggregateTally  *self,
+                        const char          *date,
+                        GError             **error)
+{
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GFile) file = NULL;
+  g_autofree gchar *path = NULL;
+
+  path = g_build_filename (self->persistent_cache_directory,
+                           "aggregate-timers",
+                           date,
+                           NULL);
+  file = g_file_new_for_path (path);
+  g_file_make_directory_with_parents (file, NULL, &local_error);
+
+  if (local_error)
+    {
+      if (g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        return;
+      g_propagate_error (error, g_steal_pointer (&local_error));
+    }
+}
+
+static inline GVariant *
+aggregate_timer_to_variant (guint32    unix_user_id,
+                            GVariant  *event_id,
+                            GVariant  *aggregate_key,
+                            GVariant  *payload,
+                            guint32    counter,
+                            gint64     monotonic_time_us)
+{
+  return g_variant_new (AGGREGATE_TIMER_VARIANT_TYPE,
+                        event_id,
+                        aggregate_key,
+                        unix_user_id,
+                        counter,
+                        payload);
+}
+
+static gchar *
+get_tally_path_from_aggregate_timer (EmerAggregateTally *self,
+                                     guint32             unix_user_id,
+                                     GVariant           *event_id,
+                                     GVariant           *aggregate_key,
+                                     const char         *date)
+{
+  g_autofree gchar *aggregate_key_str = NULL;
+  g_autofree gchar *event_id_str = NULL;
+  g_autofree gchar *filename = NULL;
+
+  aggregate_key_str = g_variant_print (aggregate_key, FALSE);
+  event_id_str = g_variant_print (event_id, FALSE);
+  filename = g_strdup_printf ("%u-%u-%u",
+                              g_str_hash (aggregate_key_str),
+                              unix_user_id,
+                              g_str_hash (event_id_str));
+
+  return g_build_filename (self->persistent_cache_directory,
+                           "aggregate-timers",
+                           date,
+                           filename,
+                           NULL);
+}
+
+static GVariant*
+load_aggregate_timer_from_path (const char *path)
+{
+  g_autoptr(GVariant) variant = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *contents = NULL;
+  gsize size;
+
+  if (!g_file_test (path, G_FILE_TEST_EXISTS))
+    return NULL;
```

Comment:
There’s a [TOCTTOU race](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) between this `g_file_test()` call and `g_file_get_contents()` below. Drop the `g_file_test()` call, and ignore the error from `g_file_get_contents()` if it’s `G_FILE_ERROR_NOENT`.